server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1583,21 +1583,34 @@ private void finalizeSnapshotEntry(Snapshot snapshot, Metadata metadata, Reposit
                     entry.startTime(),
                     indexSnapshotDetails
                 );
+                final ListenableFuture<List<ActionListener<SnapshotInfo>>> snapshotListeners = new ListenableFuture<>();
                 repo.finalizeSnapshot(
                     new FinalizeSnapshotContext(
                         shardGenerations,
                         repositoryData.getGenId(),
                         metaForSnapshot,
                         snapshotInfo,
                         entry.version(),
-                        ActionListener.wrap(
-                            updatedRepositoryData -> runNextQueuedOperation(updatedRepositoryData, repository, true),
-                            e -> handleFinalizationFailure(e, snapshot, repositoryData)
-                        ),
-                        snInfo -> {
-                            completeListenersIgnoringException(endAndGetListenersToResolve(snInfo.snapshot()), snInfo);
-                            logger.info("snapshot [{}] completed with state [{}]", snapshot, snInfo.state());
-                        }
+                        ActionListener.wrap(updatedRepositoryData -> {
+                            // get a hold of the listeners for this snapshot here and store them in the future so they can be used
+                            // by the snapshot info callback below and won't be failed needlessly if #runNextQueuedOperation runs into
+                            // a fatal like e.g. this node stopped being the master node
+                            snapshotListeners.onResponse(endAndGetListenersToResolve(snapshot));
+                            runNextQueuedOperation(updatedRepositoryData, repository, true);
+                        }, e -> handleFinalizationFailure(e, snapshot, repositoryData)),
+                        snInfo -> snapshotListeners.addListener(new ActionListener<>() {
+                            @Override
+                            public void onResponse(List<ActionListener<SnapshotInfo>> actionListeners) {
+                                completeListenersIgnoringException(actionListeners, snInfo);
+                                logger.info("snapshot [{}] completed with state [{}]", snapshot, snInfo.state());
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
```

Comment:
We have this pattern (a future that only succeeds but never fails) in a couple of spots now. I'll try to create a follow-up to dry this up, it's quite noisy doing this manually everywhere.